### PR TITLE
Fix integer overflow in exgcd

### PR DESCRIPTION
**Pull request by liangjs**
The old exgcd can overflow when its arguments are close to the max/min
of the signed integer type. So we convert the arguments to unsigned
integers before computation. This will make sure the value of ta and tb
always decrease in iterations.

This also introduces the sign and unsigned_abs utility functions.

### CODE REVIEW

**Comment by liangjs** [comment]
[Here](https://github.com/gcc-mirror/gcc/blob/16e2427f50c208dfe07d07f18009969502c25dc8/libstdc%2B%2B-v3/include/std/numeric#L96-L106) is a reference of the `unsigned_abs` implementation.

**Comment by liangjs** [comment]
[Here](https://github.com/gcc-mirror/gcc/blob/16e2427f50c208dfe07d07f18009969502c25dc8/libstdc%2B%2B-v3/include/std/numeric#L96-L106) is a reference of the `unsigned_abs` implementation.

File: number_theory/utility.h [hunk 1]
```diff
@@ -31,6 +32,38 @@ U binary_accumulate(T binary,
   return result;
 }
 
+// Returns the sign of the integer |x|.
+// If x = 0, sign(x) = 0.
+// If x > 0, sign(x) = 1.
+// If x < 0, sign(x) = -1.
+template <class T>
+constexpr int sign(T x) {
+  static_assert(std::numeric_limits<T>::is_integer,
```
Comment (umutoztunc) [inline number_theory/utility.h]: Let's be consistent and end all assertion sentences with a full-stop.

File: number_theory/utility.h [hunk 2]
```diff
@@ -31,6 +32,38 @@ U binary_accumulate(T binary,
...
+
+// Returns the absolute value of the integer |x| in an unsigned type.
+//
+// This can be useful because std::abs doesn't support unsigned integers, and
+// std::abs(std::numeric_limits<T>::min()) is undefined.
+template <class T, class U = std::make_unsigned_t<T>>
+constexpr U unsigned_abs(T x) {
+  static_assert(std::numeric_limits<T>::is_integer,
+                "unsigned_abs argument |x| must be an integer");
+  static_assert(std::is_unsigned<U>::value,
```
Comment (umutoztunc) [inline number_theory/utility.h]: I suggest removing the return type from the template so that return type is always unsigned T.

Comment (umutoztunc) [reply number_theory/utility.h]: Also, we should prefer `std::numeric_limits::is_unsigned` over `std::is_unsigned` so that we can support user-defined types.

File: number_theory/numeric_unittest.cpp [hunk 3]
```diff
@@ -41,6 +41,11 @@ TEST(NumericTest, ExGCD) {
     int64_t b = rand_gen(rand_engine);
     test_exgcd(a, b);
   }
+  // test overflow
+  test_exgcd<int32_t>(std::numeric_limits<int32_t>::min(),
```
Comment (umutoztunc) [inline number_theory/numeric_unittest.cpp]: Is there a need for explicitly testing 32-bit and 64-bit overflow? If not,  we can just use `int` as suggested by Google's style guide such as:
```cpp
test_exgcd(std::numeric_limits<int>::min(),
           std::numeric_limits<int>::max());
```

**Review by umutoztunc — CHANGES_REQUESTED** [review]

File: number_theory/numeric_unittest.cpp (hunk 3 above)
Comment (liangjs) [reply number_theory/numeric_unittest.cpp]: On some 64-bit systems, int is a 32-bit integer type. Since the system uses 64-bit registers, adding two large 32-bit integer may not overflow. So we test both 32-bit and 64-bit integers explicitly.

**Review by umutoztunc — APPROVED** [review]
Great work! LGTM.
